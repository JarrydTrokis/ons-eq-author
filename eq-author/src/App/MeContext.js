import React, { createContext, useState } from "react";
import PropType from "prop-types";
import { Query, withApollo } from "react-apollo";
import { withRouter } from "react-router";
import gql from "graphql-tag";
import { get, flowRight } from "lodash";
import auth from "components/Auth";
import config from "config";
import CustomPropTypes from "custom-prop-types";
import {
  setSentryUser,
  setSentryTag,
  sendSentryError,
} from "../apollo/sentryUtils";

const createSignIn = (setHasAccessToken, history) => user => {
  localStorage.setItem("accessToken", user.ra);
  localStorage.setItem("refreshToken", user.refreshToken);

  return window
    .fetch("/signIn", {
      method: "POST",
      headers: { authorization: `Bearer ${user.ra}` },
    })
    .then(res => {
      if (!res.ok) {
        throw Error(`Server responded with a ${res.status} code.`);
      }
      setHasAccessToken(true);
      history.push(get(history, "location.state.returnURL") || "/");
    })
    .catch(e => {
      localStorage.removeItem("accessToken");
      localStorage.removeItem("refreshToken");
      auth.signOut();
      setSentryUser(user.ra);
      setSentryTag("Signing in error");
      sendSentryError(e);
      throw e;
    });
};

const createSignOut = (setHasAccessToken, { history, client }) => () => {
  if (config.REACT_APP_FULLSTORY_ORG) {
    window.FS.identify(false);
  }
  setHasAccessToken(false);
  client.clearStore();
  auth.signOut();
  localStorage.removeItem("accessToken");
  localStorage.removeItem("refreshToken");
  history.push("/sign-in");
};

export const CURRENT_USER_QUERY = gql`
  query GetMe {
    me {
      id
      displayName
      picture
      email
    }
  }
`;

export const MeContext = createContext();

const FragmentWithChildren = ({ children }) => children({});

const ContextProvider = props => {
  const accessToken = localStorage.getItem("accessToken");
  const [hasAccessToken, setHasAccessToken] = useState(Boolean(accessToken));
  const signIn = createSignIn(setHasAccessToken, props.history);
  const signOut = createSignOut(setHasAccessToken, props);
  const QueryOrFragment = hasAccessToken ? Query : FragmentWithChildren;
  return (
    <QueryOrFragment query={CURRENT_USER_QUERY}>
      {innerProps => {
        const me = get(innerProps, "data.me");
        return (
          <MeContext.Provider
            value={{
              me,
              signIn,
              hasAccessToken,
              signOut,
              awaitingUserQuery: innerProps.loading,
            }}
          >
            {props.children}
          </MeContext.Provider>
        );
      }}
    </QueryOrFragment>
  );
};

ContextProvider.propTypes = {
  history: CustomPropTypes.history.isRequired,
  children: PropType.node,
};

export const MeProvider = flowRight(
  withApollo,
  withRouter
)(ContextProvider);

export const withMe = Component => props => (
  <MeContext.Consumer>
    {({ me, signIn, hasAccessToken, awaitingUserQuery, signOut }) => (
      <Component
        {...props}
        me={me}
        signIn={signIn}
        signOut={signOut}
        hasAccessToken={hasAccessToken}
        awaitingUserQuery={awaitingUserQuery}
      />
    )}
  </MeContext.Consumer>
);
