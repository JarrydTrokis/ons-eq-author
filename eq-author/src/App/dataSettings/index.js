import React from "react";
import { Route, Redirect } from "react-router-dom";
import SampleFileDataPage from "./SampleFileData";
import SupplementaryDataPage from "./SupplementaryDataPage";
import { buildDataPath } from "utils/UrlUtils";

export default [
  <Route
    key="sample"
    exact
    path="/q/:questionnaireId/data/sample-file-data"
    component={SampleFileDataPage}
  />,
  <Route
    key="dataset"
    exact
    path="/q/:questionnaireId/data/supplementary-data"
    component={SupplementaryDataPage}
  />,
  <Route
    key="data"
    exact
    path="/q/:questionnaireId/data"
    render={(props) => (
      <Redirect to={`${buildDataPath(props.match.params)}/sample-file-data`} />
    )}
  />,
];
