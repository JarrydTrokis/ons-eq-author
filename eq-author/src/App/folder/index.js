import React from "react";

import { Route, Redirect } from "react-router-dom";
import { Query } from "react-apollo";
import SkipLogic from "App/shared/Logic/SkipLogic";

import Design from "./Design";
import RoutingPage from "./Routing";

import Error from "components/Error";

import GET_FOLDER_QUERY from "./graphql/getFolderQuery.graphql";

export default [
  <Route
    key="folder-design"
    path="/q/:questionnaireId/folder/:folderId/design"
    component={Design}
  />,
  <Route key="folder-logic" path="/q/:questionnaireId/folder/:folderId/logic">
    <Redirect to="routing" />
  </Route>,
  <Route
    key="folder-logic-routing"
    path="/q/:questionnaireId/folder/:folderId/routing"
    render={(props) => (
      <Query
        query={GET_FOLDER_QUERY}
        variables={{
          input: { folderId: props.match.params.folderId },
        }}
      >
        {({ loading, error, data }) => {
          if (loading) {
            return <React.Fragment />;
          }

          if (error) {
            return <Error>Error fetching folder</Error>;
          }

          if (data) {
            const { folder } = data;
            return folder.listId ? (
              <Redirect to="design" />
            ) : (
              <RoutingPage {...props} />
            );
          }
        }}
      </Query>
    )}
  />,
  <Route
    key="folder-logic-skip"
    path="/q/:questionnaireId/folder/:folderId/skip"
    render={(props) => (
      <Query
        query={GET_FOLDER_QUERY}
        variables={{
          input: { folderId: props.match.params.folderId },
        }}
      >
        {({ loading, error, data }) => {
          if (loading) {
            return <React.Fragment />;
          }

          if (error) {
            return <Error>Error fetching folder</Error>;
          }

          if (data) {
            const { folder } = data;
            return folder.listId ? (
              <Redirect to="design" />
            ) : (
              <SkipLogic {...props} />
            );
          }
        }}
      </Query>
    )}
  />,
];
