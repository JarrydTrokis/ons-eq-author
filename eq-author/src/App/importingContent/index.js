import React, { useState } from "react";
import { Query } from "react-apollo";
import { useParams, useLocation } from "react-router-dom";

import * as Headings from "constants/table-headings";

import GET_QUESTIONNAIRE_LIST from "graphql/getQuestionnaireList.graphql";
import GET_QUESTIONNAIRE from "graphql/getQuestionnaire.graphql";

import QuestionnairesView from "components/QuestionnairesView";
import QuestionnaireSelectModal from "components/modals/QuestionnaireSelectModal";
import ReviewQuestionsModal from "components/modals/ImportQuestionReviewModal";
import QuestionPicker from "components/QuestionPicker";

const SelectQuestionnaire = ({
  isOpen,
  questionnaires = [],
  onSelect,
  onCancel,
}) => {
  const [selectedQuestionnaire, setSelectedQuestionnaire] = useState(null);

  const enabledHeadings = [
    Headings.TITLE,
    Headings.OWNER,
    Headings.CREATED,
    Headings.MODIFIED,
  ];

  return (
    <QuestionnaireSelectModal
      isOpen={isOpen}
      onSelect={() => onSelect(selectedQuestionnaire)}
      onClose={onCancel}
      onCancel={onCancel}
      disableSelect={!selectedQuestionnaire}
    >
      <QuestionnairesView
        questionnaires={questionnaires}
        selectedQuestionnaire={selectedQuestionnaire}
        enabledHeadings={enabledHeadings}
        canCreateQuestionnaire={false}
        padding="small"
        onQuestionnaireClick={(questionnaireId) =>
          setSelectedQuestionnaire(
            questionnaires.find(({ id }) => id === questionnaireId)
          )
        }
      />
    </QuestionnaireSelectModal>
  );
};

const ImportingContent = ({ stopImporting }) => {
  /*
   * Modal display states
   */

  //True as it's the first modal in the process
  const [selectingQuestionnaire, setSelectingQuestionnaire] = useState(true);
  const [reviewingQuestions, setReviewingQuestions] = useState(false);
  const [selectingQuestions, setSelectingQuestions] = useState(false);

  /*
   * Data
   */

  const [questionnaireImportingFrom, setQuestionnaireImportingFrom] =
    useState(null);

  const [questionsToImport, setQuestionsToImport] = useState([]);

  /*
   * Handlers
   */

  const onSelectQuestionnaire = (questionnaire) => {
    setQuestionnaireImportingFrom(questionnaire);
    setSelectingQuestionnaire(false);
    setReviewingQuestions(true);
  };

  const onSelectQuestions = () => {
    setReviewingQuestions(false);
    setSelectingQuestions(true);
  };

  const onBackFromReviewingQuestions = () => {
    setReviewingQuestions(false);
    setSelectingQuestionnaire(true);
  };

  const onQuestionPickerCancel = () => {
    setSelectingQuestions(false);
    setReviewingQuestions(true);
  };

  const onQuestionPickerSubmit = (selection) => {
    setQuestionsToImport(selection);
    setSelectingQuestions(false);
    setReviewingQuestions(true);
  };

  const onGlobalCancel = () => {
    setReviewingQuestions(false);
    setSelectingQuestionnaire(false);
    setQuestionnaireImportingFrom(null);
    stopImporting();
  };

  const { questionnaireId: currentQuestionnaireId } = useParams();

  return (
    <>
      {selectingQuestionnaire && (
        <Query
          query={GET_QUESTIONNAIRE_LIST}
          variables={{
            input: { filter: { ne: { ids: [currentQuestionnaireId] } } },
          }}
        >
          {({ loading, error, data }) => {
            if (loading) {
              return <p>Loading</p>;
            }

            if (error || !data) {
              return <p>Error</p>;
            }

            return (
              <SelectQuestionnaire
                isOpen={selectingQuestionnaire}
                onCancel={onGlobalCancel}
                questionnaires={data.questionnaires}
                onSelect={onSelectQuestionnaire}
              />
            );
          }}
        </Query>
      )}
      {reviewingQuestions && (
        <ReviewQuestionsModal
          isOpen={reviewingQuestions}
          questionnaire={questionnaireImportingFrom}
          onCancel={onGlobalCancel}
          onBack={onBackFromReviewingQuestions}
          onSelectQuestions={onSelectQuestions}
          startingSelectedQuestions={questionsToImport}
        />
      )}
      {selectingQuestions && (
        <Query
          query={GET_QUESTIONNAIRE}
          variables={{
            input: { questionnaireId: questionnaireImportingFrom.id },
          }}
        >
          {({ loading, error, data }) => {
            if (loading) {
              return <p>Loading</p>;
            }

            if (error || !data) {
              return <p>Error</p>;
            }

            const { sections } = data.questionnaire;

            return (
              <QuestionPicker
                isOpen={selectingQuestions}
                sections={sections}
                startingSelectedQuestions={questionsToImport}
                warningPanel="You cannot import folders but you can import any questions they contain."
                showSearch
                onClose={onGlobalCancel}
                onCancel={onQuestionPickerCancel}
                onSubmit={onQuestionPickerSubmit}
              />
            );
          }}
        </Query>
      )}
    </>
  );
};

export default ImportingContent;
