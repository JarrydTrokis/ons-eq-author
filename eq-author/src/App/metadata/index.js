import React from "react";

import RedirectRoute from "components/RedirectRoute";

export default [
  <RedirectRoute
    key="metadata"
    from="/q/:questionnaireId/metadata"
    to="/q/:questionnaireId/data/sample-file-data"
  />,
];
