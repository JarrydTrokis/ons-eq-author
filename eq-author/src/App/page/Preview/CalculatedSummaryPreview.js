/* eslint-disable react/no-danger */
import React from "react";
import { propType } from "graphql-anywhere";

import styled from "styled-components";
import Error from "components/preview/Error";
import PageTitle from "components/preview/elements/PageTitle";
import Info from "components/preview/elements/Info";

import EditorLayout from "components/EditorLayout";
import Panel from "components/Panel";
import { Grid, Column } from "components/Grid";
import CommentsPanel from "App/Comments";

import { colors } from "constants/theme";
import CalculatedSummaryPageEditor from "../Design/CalculatedSummaryPageEditor";
import { useSetNavigationCallbacksForPage } from "components/NavigationCallbacks";

const Button = styled.div`
  color: white;
  background-color: ${colors.positive};
  border: 2px solid ${colors.positive};
  padding: 0.75rem 1rem;
  margin: 0;
  font-size: 1rem;
  font-weight: 600;
  border-radius: 3px;
  display: inline-block;
  text-rendering: optimizeLegibility;
  margin-bottom: 2em;
`;

const Container = styled.div`
  padding: 2em;
  font-size: 1.1em;
  p {
    margin: 0 0 1em;
  }
  p:last-of-type {
    margin-bottom: 0;
  }
  em {
    background-color: #dce5b0;
    padding: 0 0.125em;
    font-style: normal;
  }
  span[data-piped] {
    background-color: #e0e0e0;
    padding: 0 0.125em;
    border-radius: 4px;
    white-space: pre-wrap;
  }
`;

const Summary = styled.div`
  border-bottom: 1px solid #999;
  margin-bottom: 2rem;
`;

const SummaryItem = styled.div`
  border-top: 1px solid #999;
  border-radius: 0;
  position: relative;
  padding: 1rem 0;
`;

const SummaryLabel = styled.div`
  font-weight: normal;
  word-break: break-word;
`;

const SummaryValue = styled.div`
  background: #e0e0e0;
  color: ${colors.textLight};
  line-height: 1;
  padding: 0.2em 0.5em;
  text-align: center;
  font-size: 0.8em;
  font-weight: bold;
  display: inline;
  border-radius: 4px;
`;

const SummaryLink = styled.div`
  color: ${colors.primary};
  text-align: right;
  text-decoration: underline;
`;

const SummaryTotal = styled(SummaryItem)`
  font-size: 1.2em;
`;

const SummaryTotalLabel = styled.div`
  font-weight: bold;
  word-break: break-word;
`;

const CalculatedSummaryPagePreview = ({ page }) => {
  useSetNavigationCallbacksForPage({
    page: page,
    folder: page?.folder,
    section: page?.section,
  });

  return (
    <EditorLayout
      title={page.displayName}
      preview
      logic
      validationErrorInfo={page.validationErrorInfo}
      renderPanel={() => (
        <CommentsPanel comments={page.comments} componentId={page.id} />
      )}
      comments={page.comments}
    >
      <Panel data-test="calSum test page">
        <Container>
          <PageTitle title={page.title} />
          {page.summaryAnswers.length > 0 ? (
            <Summary>
              {page.summaryAnswers.map((answer) => (
                <SummaryItem key={answer.id}>
                  <Grid>
                    <Column cols={7}>
                      <SummaryLabel data-test="answer-item">
                        {answer.displayName}
                      </SummaryLabel>
                    </Column>
                    <Column cols={3}>
                      <SummaryValue>Value</SummaryValue>
                    </Column>
                    <Column cols={2}>
                      <SummaryLink>Change</SummaryLink>
                    </Column>
                  </Grid>
                </SummaryItem>
              ))}

              {page.totalTitle ? (
                <SummaryTotal>
                  <Grid>
                    <Column cols={7}>
                      <SummaryTotalLabel
                        data-test="total-title"
                        dangerouslySetInnerHTML={{ __html: page.totalTitle }}
                      />
                    </Column>
                    <Column cols={3}>
                      <SummaryValue>Value</SummaryValue>
                    </Column>
                    <Column cols={2} />
                  </Grid>
                </SummaryTotal>
              ) : (
                <Error large data-test="no-total-title">
                  Missing total title
                </Error>
              )}
            </Summary>
          ) : (
            <Error large data-test="no-answers-selected">
              No answers selected
            </Error>
          )}
          <Button>Yes, I confirm this is correct</Button>
        </Container>
      </Panel>
    </EditorLayout>
  );
};

CalculatedSummaryPagePreview.propTypes = {
  page: propType(CalculatedSummaryPageEditor.fragments.CalculatedSummaryPage),
};

export default CalculatedSummaryPagePreview;
