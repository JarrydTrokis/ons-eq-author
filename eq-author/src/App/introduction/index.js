import React from "react";

import { Route } from "react-router-dom";

import Design from "./Design";
import Preview from "./Preview";

export default [
  <Route
    key="page-design"
    path="/q/:questionnaireId/introduction/:introductionId/design"
    component={Design}
  />,
  <Route
    key="page-design"
    path="/q/:questionnaireId/introduction/:introductionId/preview"
    component={Preview}
  />,
];
