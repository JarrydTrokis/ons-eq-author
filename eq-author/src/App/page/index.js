import React from "react";

import { Route, Redirect } from "react-router-dom";

import { Query } from "react-apollo";

import Design from "./Design";
import Preview from "./Preview";
import Routing from "./Logic/Routing";
import SkipLogic from "App/shared/Logic/SkipLogic";
import Error from "components/Error";

import isListCollectorPageType from "utils/isListCollectorPageType";

import GET_PAGE_QUERY from "graphql/getPage.graphql";

export default [
  <Route
    key="page-design"
    path="/q/:questionnaireId/page/:pageId/design/:modifier?"
    component={Design}
  />,
  <Route
    key="page-preview"
    path="/q/:questionnaireId/page/:pageId/preview"
    component={Preview}
  />,
  <Route key="page-logic" path="/q/:questionnaireId/page/:pageId/logic">
    <Redirect to="routing" />
  </Route>,
  <Route
    key="page-logic-routing"
    path="/q/:questionnaireId/page/:pageId/routing"
    component={Routing}
  />,
  <Route
    key="page-logic-skip"
    path="/q/:questionnaireId/page/:pageId/skip"
    render={(props) => (
      <Query
        query={GET_PAGE_QUERY}
        variables={{
          input: { pageId: props.match.params.pageId },
        }}
      >
        {({ loading, error, data }) => {
          if (loading) {
            return <React.Fragment />;
          }

          if (error) {
            return <Error>Error fetching page</Error>;
          }

          if (data) {
            const { page } = data;
            return isListCollectorPageType(page.pageType) ? (
              <Redirect to="design" />
            ) : (
              <SkipLogic {...props} />
            );
          }
        }}
      </Query>
    )}
  />,
];
