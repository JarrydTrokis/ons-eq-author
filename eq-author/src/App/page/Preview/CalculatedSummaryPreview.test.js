import React from "react";
import { shallow } from "enzyme";
import { render, flushPromises, act } from "tests/utils/rtl";

import commentsSubscription from "graphql/subscriptions/commentSubscription.graphql";
import { MeContext } from "App/MeContext";

import { byTestAttr } from "tests/utils/selectors";

import CalculatedSummaryPreview from "./CalculatedSummaryPreview";
import { publishStatusSubscription } from "components/EditorLayout/Header";
import QuestionnaireContext from "components/QuestionnaireContext";

const questionnaire = {
  id: "questionnaire-1",
  sections: [
    {
      id: "section-1",
      folders: [
        {
          id: "folder-1",
          pages: [
            {
              id: "3",
              pageType: "QuestionPage",
              title: "Page 1",
              answers: [
                { id: "4", label: "answer 1", questionPageId: "123" },
                {
                  id: "5",
                  label: "answer 2",
                  questionPageId: "234",
                },
              ],
            },
            {
              id: "6",
              pageType: "QuestionPage",
              title: "Page 2",
              answers: [
                {
                  id: "7",
                  label: "answer 3",
                  questionPageId: "345",
                },
                {
                  id: "8",
                  label: "answer 4",
                  questionPageId: "456",
                },
              ],
            },
            {
              id: "9",
              title: "calculated summary page 1",
              pageType: "CalculatedSummaryPage",
              summaryAnswers: [
                { id: "8", displayName: "Answer 8" },
                { id: "5", displayName: "Answer 5" },
                { id: "7", displayName: "Answer 7" },
                { id: "4", displayName: "Answer 4" },
              ],
            },
          ],
        },
      ],
    },
  ],
};

describe("CalculatedSummaryPreview", () => {
  let page, me, mocks, questionnaireId;

  beforeEach(() => {
    questionnaireId = "111";
    me = {
      id: "123",
      displayName: "Raymond Holt",
      email: "RaymondHolt@TheNineNine.com",
      picture: "http://img.com/avatar.jpg",
      admin: true,
    };

    page = {
      id: "pageId",
      displayName: "Question",
      position: 1,
      title: "<p>Hello world</p>",
      totalTitle: "<p>Total be:</p>",
      pageDescription: "Hello",
      alias: "Who am I?",
      type: "Number",
      answers: [],
      comments: [],
      summaryAnswers: [
        { id: "1", displayName: "Answer 1" },
        { id: "2", displayName: "Answer 2" },
        { id: "3", displayName: "Answer 3" },
      ],
      pageType: "CalculatedSummaryPage",
      section: {
        id: "1",
        position: 0,
        questionnaire: {
          id: "1",
          metadata: [],
        },
      },
      validationErrorInfo: {
        id: "test",
        totalCount: 0,
        errors: [],
      },
      folder: {
        id: "folder-1",
        position: 0,
      },
    };

    mocks = [
      {
        request: {
          query: publishStatusSubscription,
          variables: { id: questionnaireId },
        },
        result: () => ({
          data: {
            publishStatusUpdated: {
              id: questionnaireId,
              publishStatus: "Unpublished",
              __typename: "Questionnaire",
            },
          },
        }),
      },
      {
        request: {
          query: commentsSubscription,
          variables: { id: "pageId" },
        },
        result: () => {
          return {};
        },
      },
    ];
  });

  it("should render", async () => {
    const { getByTestId, getByText } = render(
      <MeContext.Provider value={{ me }}>
        <CalculatedSummaryPreview page={page} />
      </MeContext.Provider>,
      {
        route: `/q/${questionnaireId}/page/2/preview`,
        urlParamMatcher: "/q/:questionnaireId/page/:pageId",
        mocks,
      }
    );
    await act(async () => {
      await flushPromises();
    });
    expect(getByTestId("calSum test page")).toBeTruthy();
    expect(getByText("Answer 1")).toBeTruthy();
  });

  it("should render empty box when no total-title given", () => {
    page.totalTitle = "";
    const wrapper = shallow(
      <MeContext.Provider value={{ me }}>
        <CalculatedSummaryPreview page={page} />
      </MeContext.Provider>,
      {
        route: `/q/${questionnaireId}/page/2/preview`,
        urlParamMatcher: "/q/:questionnaireId/page/:pageId",
        mocks,
      }
    );
    expect(wrapper.find(byTestAttr("no-total-title"))).toBeTruthy();
  });

  it("should render 'no answers selected' message", () => {
    page.summaryAnswers = [];
    const wrapper = shallow(
      <MeContext.Provider value={{ me }}>
        <CalculatedSummaryPreview page={page} />
      </MeContext.Provider>,
      {
        route: `/q/${questionnaireId}/page/2/preview`,
        urlParamMatcher: "/q/:questionnaireId/page/:pageId",
        mocks,
      }
    );
    expect(wrapper.find(byTestAttr("no-answers-selected"))).toBeTruthy();
  });

  // test the answer.display name is rendered in the page.
  // change the code where it test based on the data-test id(data-test id need to be unique for each answer) no functions need to be mock at all.
  // check the text of answer found from the data test id is equal to the answer.displayName

  it("should order the summary answers correctly ", () => {
    const { getByTestId, getByText } = render(
      <QuestionnaireContext.Provider value={{ questionnaire }}>
        <MeContext.Provider value={{ me }}>
          <CalculatedSummaryPreview page={page} />
        </MeContext.Provider>
      </QuestionnaireContext.Provider>,
      {
        route: `/q/${questionnaireId}/page/9/preview`,
        urlParamMatcher: "/q/:questionnaireId/page/:pageId",
        mocks,
      }
    );
    questionnaire.sections[0].folders[0].pages[2].summaryAnswers.forEach(
      (answer) => {
        const renderDisplayName = getByText(answer.displayName);
        expect(renderDisplayName).toBeInTheDocument();
      }
    );
  });
});
