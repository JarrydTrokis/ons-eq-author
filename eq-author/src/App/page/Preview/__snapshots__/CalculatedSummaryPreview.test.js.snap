// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatedSummaryPreview should render 1`] = `
<EditorLayout
  page={
    Object {
      "alias": "Who am I?",
      "displayName": "Question",
      "id": "1",
      "pageType": "CalculatedSummaryPage",
      "position": 1,
      "section": Object {
        "id": "1",
        "questionnaire": Object {
          "id": "1",
          "metadata": Array [],
        },
      },
      "summaryAnswers": Array [
        Object {
          "displayName": "Answer 1",
          "id": "1",
        },
        Object {
          "displayName": "Answer 2",
          "id": "2",
        },
        Object {
          "displayName": "Answer 3",
          "id": "3",
        },
      ],
      "title": "<p>Hello world</p>",
      "totalTitle": "<p>Total be:</p>",
    }
  }
  preview={true}
  routing={false}
>
  <Panel>
    <CalculatedSummaryPreview__Container>
      <PageTitle
        title="<p>Hello world</p>"
      />
      <Info>
        Please review your answers and confirm these are correct.
      </Info>
      <CalculatedSummaryPreview__Summary>
        <CalculatedSummaryPreview__SummaryItem
          key="1"
        >
          <Grid
            align="top"
            fillHeight={true}
          >
            <Column
              cols={7}
            >
              <CalculatedSummaryPreview__SummaryLabel
                data-test="answer-item"
              >
                Answer 1
              </CalculatedSummaryPreview__SummaryLabel>
            </Column>
            <Column
              cols={3}
            >
              <CalculatedSummaryPreview__SummaryValue>
                Value
              </CalculatedSummaryPreview__SummaryValue>
            </Column>
            <Column
              cols={2}
            >
              <CalculatedSummaryPreview__SummaryLink>
                Change
              </CalculatedSummaryPreview__SummaryLink>
            </Column>
          </Grid>
        </CalculatedSummaryPreview__SummaryItem>
        <CalculatedSummaryPreview__SummaryItem
          key="2"
        >
          <Grid
            align="top"
            fillHeight={true}
          >
            <Column
              cols={7}
            >
              <CalculatedSummaryPreview__SummaryLabel
                data-test="answer-item"
              >
                Answer 2
              </CalculatedSummaryPreview__SummaryLabel>
            </Column>
            <Column
              cols={3}
            >
              <CalculatedSummaryPreview__SummaryValue>
                Value
              </CalculatedSummaryPreview__SummaryValue>
            </Column>
            <Column
              cols={2}
            >
              <CalculatedSummaryPreview__SummaryLink>
                Change
              </CalculatedSummaryPreview__SummaryLink>
            </Column>
          </Grid>
        </CalculatedSummaryPreview__SummaryItem>
        <CalculatedSummaryPreview__SummaryItem
          key="3"
        >
          <Grid
            align="top"
            fillHeight={true}
          >
            <Column
              cols={7}
            >
              <CalculatedSummaryPreview__SummaryLabel
                data-test="answer-item"
              >
                Answer 3
              </CalculatedSummaryPreview__SummaryLabel>
            </Column>
            <Column
              cols={3}
            >
              <CalculatedSummaryPreview__SummaryValue>
                Value
              </CalculatedSummaryPreview__SummaryValue>
            </Column>
            <Column
              cols={2}
            >
              <CalculatedSummaryPreview__SummaryLink>
                Change
              </CalculatedSummaryPreview__SummaryLink>
            </Column>
          </Grid>
        </CalculatedSummaryPreview__SummaryItem>
        <CalculatedSummaryPreview__SummaryTotal>
          <Grid
            align="top"
            fillHeight={true}
          >
            <Column
              cols={7}
            >
              <CalculatedSummaryPreview__SummaryTotalLabel
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "<p>Total be:</p>",
                  }
                }
                data-test="total-title"
              />
            </Column>
            <Column
              cols={3}
            >
              <CalculatedSummaryPreview__SummaryValue>
                Value
              </CalculatedSummaryPreview__SummaryValue>
            </Column>
            <Column
              cols={2}
            />
          </Grid>
        </CalculatedSummaryPreview__SummaryTotal>
      </CalculatedSummaryPreview__Summary>
    </CalculatedSummaryPreview__Container>
  </Panel>
</EditorLayout>
`;
